### PR TITLE
Hoist await and yield

### PR DESCRIPTION
**Pull request by novemberborn**
General code cleanup and dependency upgrades, as well as a fix for https://github.com/avajs/ava/issues/987 (see last commit).

The `await` and `yield` expressions may still throw of course. I'm wondering if we want to capture those exceptions with this helper? They wouldn't be due to misuse of `t.throws()` however.

### CODE REVIEW

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Looks good :)

@jfmengels Could you review https://github.com/avajs/babel-plugin-ava-throws-helper/pull/6/commits/0f850a2aa2eef667ad40f7d5fcbeeaf2b1ecf7fd ?

File: index.js [hunk 1]
```diff
@@ -40,14 +42,24 @@ var assertionVisitor = {
...
 				LINE: t.numericLiteral(arg0.loc.start.line),
 				COLUMN: t.numericLiteral(arg0.loc.start.column),
 				SOURCE: t.stringLiteral(state.file.code.substring(arg0.start, arg0.end)),
 				FILE: t.stringLiteral(state.file.opts.filename)
 			}).expression;
+
+			if (hoisted) {
+				path.replaceWithMultiple(
+					hoisted.expressions.concat(
+						t.expressionStatement(path.node)
```
Comment (jfmengels) [inline index.js:59]: You are replacing the Callxpression by an ExpressionStatement here. Would the following test code stll work, or would Babel crash or something because you are building an odd AST?

``` js
test('should foo...', async t => {
  t.is(
      t.throws(foo(await quux())).message,
     'Could not find X'
   );
});
```

My worry is that you can't AFAIK pass a statement as the argument of a function (here, the `t.is(){ call). But I don't know how Babel would react to that. Would you mind adding a test case and see how that pans out?
I would be fine with not entirely supporting this edge case, but it should at the very least not crash.

**Review by jfmengels — CHANGES_REQUESTED** [review]
This looks great. And pretty nice to read too (haven't looked at Babel code in a while).

I just have one concern for which I'd like to see some tests, but otherwise, LGTM :)

File: index.js (hunk 1 above)
Comment (novemberborn) [reply index.js:59]: Yea I doubt that'll work… in your example `t.throws` is inside an `ArgumentExpression` so we'd have to walk the path back to the test implementation and detect any `ArgumentExpression` nodes, then prepend the hosted statements to node's parent.

But then what about:

``` js
test(async t => {
  (function* () {
    const err = await t.throws(foo(yield))
    t.true(err.message === 'hahaha')
  })()
})
```

That scenario is further complicated by this plugin running _after_ async/await has been transpiled into a generator…

Comment (sindresorhus) [reply index.js:59]: So, what should we do about it? Just ignore it as an edge-case?

Comment (novemberborn) [reply index.js:59]: I haven't had the time to get back to this. We could explore whether we can run this plugin _before_ any other presets and plugins. Alternatively the generator compilation (for Node.js v4 and up) should reduce to `yield` expressions.

Comment (sindresorhus) [reply index.js:59]: Relevant: https://github.com/babel/babel/pull/3281

---

> Alternatively the generator compilation (for Node.js v4 and up) should reduce to yield expressions.

I would be ok with only supporting Node.js 4 for this.

**Comment by novemberborn** [comment]
Another approach may be to rewrite `t.throws` to something like:

```js
t.throws(
  (
    t._throwsArgStart(),
    t._throwsArgEnd(fn(await foo))
  ),
  TypeError
)
```

`_throwsArgStart()` could set a flag, which is unset by `_throwsArgEnd()`. This function returns its first argument. If an exception occurs while the flag is set (caused by the `fn(await foo)` invocation) then that's because of improper use. Otherwise the promise or callback is passed to `t.throws()` and the assertion runs as normal.

**Comment by sindresorhus** [comment]
@novemberborn I like that approach. I can't really think of any immediate downsides of doing that.

@jfmengels Thoughts? ⬆️

**Comment by novemberborn** [comment]
Closing in favor of #8.
